### PR TITLE
feat: Implemented a new vacation request cancellation feature

### PR DESCRIPTION
**Pull request by Mahmoud-Emad**
### Changes

- Updated the `STATUS_CHOICES` class.
- Created an endpoint to update the status of the request to `request_to_cancel` status.
- Added the required validation on the view handler.
- Updated the UI to display the button the cancelation only If the requested user is the applier, and the status is still pending.
- Pushed a notification for the leaders to notify them by the request.
- Working on supporting the real-time request updater.
- Renamed the deleteVacation to cancelVacation.
- Updated the behavior of request deletion to only change the status of the request to 'cancel' to keep the record in the database.
- Notify the applier and the leaders if the admin canceled the request.
- Make a condition on the cancelation to check if the status is pending.
- Removed the cancel request from the calendar.
- Implemented a new endpoint to reject the request to cancel the vacation request.
- Push notification for the applier.
- Updated the front end with the required actions.
- Working on pushing the update to the other user.
- Working on wrapping the request status.
- Check if the requested user can see the notification content.
- Added a warning modal to tell the user that 'You are not allowed to access this content'.
- Updated the request type to be 'vacations' as the model instead of 'vacation'.
- Support request to cancel the vacation if it's already approved.
- Push notifications in time to/from users.
- Formatted the code.
- Refactor the notifications consumer functionality.
- Update the calendar with the actual status of the request.

### Related Issues

- https://github.com/codescalers/CSHR/issues/430
- https://github.com/codescalers/CSHR/issues/247

### Checklist

- [ ] Tests included
- [ ] Build pass
- [ ] Documentation
- [ ] Code format and docstrings
- [ ] Screenshots/Video attached (needed for UI changes)

### CODE REVIEW

File: client/src/clients/api/vacations.ts [hunk 1]
```diff
@@ -57,10 +57,15 @@ export class VacationsApi extends ApiClientBase {
...
   }
-  async read(id: number) {}
 
-  async delete(id: number) {
-    return this.unwrap(() => this.$http.delete<any>(this.getUrl(`/${id}`)), {
+  async cancel(id: number) {
+    return this.unwrap(() => this.$http.put<any>(this.getUrl(`/cancel/${id}`)), {
+      transform: (d) => d.results
+    })
+  }
```
Comment (zaelgohary) [inline client/src/clients/api/vacations.ts:65]: All user roles should be able to cancel their pending request.

File: client/src/clients/api/vacations.ts [hunk 2]
```diff
@@ -57,10 +57,15 @@ export class VacationsApi extends ApiClientBase {
...
+
+  async requestToCancel(id: number) {
+    return this.unwrap(() => this.$http.put<any>(this.getUrl(`/request-to-cancel/${id}`)), {
       transform: (d) => d.results
     })
   }
```
Comment (zaelgohary) [inline client/src/clients/api/vacations.ts:71]: Whenever user tries to cancel his approved vacation he gets this:
You do not have permission to perform this action.

**Review by zaelgohary — CHANGES_REQUESTED** [review]
Nice work, @Mahmoud-Emad! just some enhanements that need to be taken into consideration.

- Request should have final action once approval user approve/cancel. (no need for user intervention)
- Approval user should be able to take approve/cancel from the notification
- Approval user data  in notification is incorrect

**Comment by Mahmoud-Emad** [comment]
> Nice work, @Mahmoud-Emad! just some enhanements that need to be taken into consideration.
> 
> * Request should have final action once approval user approve/cancel. (no need for user intervention)
> * Approval user should be able to take approve/cancel from the notification
> * Approval user data  in notification is incorrect

All cases covered!

**Review by zaelgohary — APPROVED** [review]

**Pull request merged**